[CBRD-25464] loaddb with pl/csql sp

hyunikn opened the pull request:
http://jira.cubrid.org/browse/CBRD-25464

- loaddb 중에 parser->original_buffer, g_query_string 이 NULL 이 됨에 따라 g_plcsql_text 가 유효하지 않은 메모리 주소를 갖게 되는 문제를 수정하였습니다. 
  - g_query_string 이 NULL 일 때는 가변 길이 버퍼를 할당하여 거기에 plcsql text 를 누적하도록 수정
- loaddb 를 실행할 때는 parser_yyinput_single_mode 를 true 로 하여 세미콜론을 만나는 데 까지만 (한번에 한 statement 만) 
  파싱하도록 합니다. 이 세미콜론 인지 과정이 종전에는 lexer 의 버퍼 채워주는 함수 YY_INPUT 에서 이루어 졌으나,
  이제는 PL/CSQL 프로그램 바디 바깥의 세미콜론에 대해서만 파싱 중지 신호를 보낼 수 있도록 Lexer 의 action으로 
  세미콜론 인지 과정을 이동했습니다. Lexer 의 상태에는 현재 PL/CSQL 프로그램 바디 안쪽인지 정보가 있습니다. 
  (mode PLCSQL_TEXT)
- 현재 다른 이유로 loaddb의 client-server 모드에서만 동작합니다. 테스트 하실 때 loaddb에 -C 옵션을 주시기 바랍니다.

hyunikn committed:
resolving core dump caused by null g_query_string and hence invalid g_plcsql_text

hyunikn committed:
draft implementation

hyunikn committed:
using only parser_yyinput_multi_line

hyunikn committed:
use statement->sql_user_text if it is avaliable

hyunikn committed:
replace ';' with the new token SEMICOLON

hyunikn committed:
Merge branch 'feature/plcsql-p1n' into cbrd-25464-loaddb-with-plcsql

hyunikn committed:
code refining: remove temporary or unused code

hyunikn committed:
code style

hgryoo commented:
test_medium CI에서 다수의 -493 에러가 발생하고 있습니다. 확인 부탁드립니다.

ctshim commented on src/parser/csql_lexer.l:
```diff
@@ -1967,95 +1980,9 @@ yywrap ()
-}
-
-static int
-parser_yyinput_single_line (char *buff, int max_size)
```
43라인에서 아래와 같이 선언된 내용도 함께 제거 될 필요가 있습다.
static int parser_yyinput_single_line (char *buff, int max_size);

hyunikn replied:
네. 감사합니다.

hyunikn committed:
Merge branch 'feature/plcsql-p1n' into cbrd-25464-loaddb-with-plcsql

hyunikn commented:
> test_medium CI에서 다수의 -493 에러가 발생하고 있습니다. 확인 부탁드립니다.

넵. 로컬에서 medium 테스트를 돌려보니 
변경된 코드에서만 어떤 assert fail 이 발생함을 확인하였습니다. (csql_grammar.y : 1834)
원인에 대해서는 좀 더 파악이 필요합니다. 

Update:

문제 부분을 수정하는 새로운 commit 을 push 했습니다. 

직전 commit 에서는 loaddb 수행 시 한 문장 수행 후에도 lexer buffer 에 다음 문장 text 가 남아 있게 됩니다. 
어떤 종류의 문장 execute 과정에서 내부적으로 SQL 문을 수행하는 경우가 있는데 
(예: call add_member ... 를 실행했을 때 object/authenticate_access_user.cpp 의 au_compute_groups() )
이 경우 실행되어야 할 SQL 문이 아니라 버퍼에 남아 있던 다음 실행문부터 파싱이 시작되는 문제가 있었습니다. 
그래서 실행할 sql 문 길이보다 buffer_pos 가 작거나 같아야 한다는 assertion fail 이 발생했었습니다. 

지웠던 parser_yyinput_single_line() 을 복원하되, 세미콜론을 만나면 -1 을 버퍼에 넣지 않고 그냥 리턴만 하도록 
수정하여 lexer 버퍼에 다음 실행문의 text 가 오지 않도록 하였습니다.

hyunikn committed:
restored parser_yyinput_single_line() with a modification
  - Now, it does not put -1 (end of file marker) to the lexer buffer at a
    semicolon, but just stops filling the buffer and returns.

ctshim commented on src/parser/csql_grammar.y:
```diff
@@ -12805,8 +12814,16 @@ pl_language_spec
+
 			  }
 
+                        if (!g_query_string) {
```
실제 free 해야 할 변수를 체크하는 것이 좋지 않을까요?
```suggestion
                        if (g_plcsql_text) {
```

ctshim commented on src/parser/csql_lexer.l:
```diff
@@ -1983,75 +1994,75 @@ parser_yyinput_single_line (char *buff, int max_size)
-	  return i;
-	}
+       {
+         buff[i] = '\0'; /* not required but to ease debugging */
```
buff[i] = '\0'; /* not required but to ease debugging */

위 내용은 2065 라인에도 추가해 주면 좋을 듯 합니다.

hyunikn replied:
네. 반영하겠습니다.

beyondykk9 commented on src/sp/jsp_cl.cpp:
```diff
@@ -841,7 +841,16 @@ jsp_create_stored_procedure (PARSER_CONTEXT *parser, PT_NODE *statement)
+      else
+	{
+	  // In particular, statement->sql_user_text is NULL when this process is from loaddb.
+	  char *plcsql_text = parser_print_tree (parser, statement);
```
custom print 플래그들을 설정하지 않아도 되나요?

hyunikn replied:
parser_print_tree 를 사용하지 않도록 재구현되었습니다.

hgryoo commented on src/sp/jsp_cl.cpp:
```diff
@@ -841,7 +841,16 @@ jsp_create_stored_procedure (PARSER_CONTEXT *parser, PT_NODE *statement)
-      compile_request.code.assign (statement->sql_user_text, statement->sql_user_text_len);
+      if (statement->sql_user_text && statement->sql_user_text_len > 0)
+	{
+	  compile_request.code.assign (statement->sql_user_text, statement->sql_user_text_len);
```
```suggestion
	  const char *plcsql_text = parser_print_tree (parser, statement);
	  compile_request.code.assign (plcsql_text, strlen (plcsql_text));
```

오프라인 리뷰에서와 같이 위와 같이 하나의 로직으로만 코드가 저장되면 좋을 것 같습니다.
이 과정으로 unloaddb에서도 다른 객체와 일관성을 맞출 수 있을 것 같습니다.

hyunikn replied:
에러 라인 수에 변경이 있을 수 있어서 sql_user_text (사용자 입력 그대로) 가 loaddb 때도 non-null 이 되도록 다시 구현했습니다.

hyunikn committed:
temp

hyunikn committed:
Merge branch 'feature/plcsql-p1n' into cbrd-25464-loaddb-with-plcsql

hyunikn committed:
[minor] simplify pl_language_spec rule: remove opt_identifier at the end

hyunikn committed:
get sql user text from the file when loaddb schema loading is being done and hence parser->original_buffer is unavailable
  . removed old code to accumulate plcsql text parts when parser->original_buffer is unavailable.

hyunikn committed:
code style

hyunikn commented:
parser->original_buffer 가 없을 때 (특히, loaddb의 schema loading 중에)
lexer token 텍스트를 누적하여 plcsql text body 를 구하던 방식을 버리고
parser->file 로부터 텍스트를 읽어 들이는 방식으로 다시 구현했습니다.  
(춘택님이 아이디어 및 초기 구현을 제공하셨습니다.)

다시 리뷰 부탁드립니다. 
주로 24558c4 commit 입니다. 

test_medium 에서 사용하는 mdb_schema 파일에 수동으로 다음과 같은 create 문을 추가하여 테스트 했습니다. 
```
create or replace procedure poo(
  i int,
  j int) as
  k int;
  l int;
begin
  null;
end;

create or replace function foo(
  i int,
  j int) return int as
  k int;
  l int;
begin
  return null;
end;

    create or replace procedure poo1(
      i int,
      j int) as
      k int;
      l int;
    begin
      null;
    end;

    create or replace function foo1(
      i int,
      j int) return int as
      k int;
      l int;
    begin
      return null;
    end;
```
mdb_schema 파일을 수정한 후
```
cubrid loaddb -C -s mdb_schema -u dba --no-user-specified-name mdb
```
명령으로 추가하니 잘 들어갑니다. 

참고로, mdb_schema 파일은 cubrid-testcases/medium/files/mdb.tar.gz 을 풀면 나옵니다.

hyunikn commented on src/parser/csql_grammar.y:
```diff
@@ -12784,26 +12814,20 @@ opt_lang_plcsql
 
 pl_language_spec
-	: opt_lang_plcsql plcsql_text opt_identifier
+	: opt_lang_plcsql plcsql_text
```
opt_identifier 를 없애는 코드 정리도 함께 했습니다. 
create procedure poo ... end poo; 와 같이 SP 이름을 end 뒤에 선택적으로 붙일 수 있는데, 
이전에는 이 이름이 plcsql_text 에 포함되지 않아 약간 지저분한 코드가 parser 쪽에 포함되었었습니다.

hyunikn commented on src/parser/csql_lexer.l:
```diff
@@ -1712,10 +1726,23 @@ _[uU][tT][fF]8[']		{
+ /* mode POST_PLCSQL_TEXT */
 
-%%
+<POST_PLCSQL_TEXT>[cC][oO][mM][mM][eE][nN][tT]          { BEGIN(INITIAL);
```
이하 변경은 CREATE ... END 뒤에 붙는 SP 이름을 PLCSQL_TEXT_SOME 토큰으로 보내기 위한 추가 구현입니다. 
이 PR 의 주 변경과 관계 없는 코드 정리에 해당됩니다.

hyunikn commented on src/sp/jsp_cl.cpp:
```diff
@@ -884,6 +884,7 @@ jsp_create_stored_procedure (PARSER_CONTEXT *parser, PT_NODE *statement)
 
   if (sp_info.lang == SP_LANG_PLCSQL)
     {
+      assert (statement->sql_user_text && statement->sql_user_text_len);
```
loaddb 일 때에도 sql_user_text 가 non-null 이 되도록 했습니다. 참조: pt_set_plcsql_body_impl()

ctshim replied:
loaddb뿐만 아니라 `csql -i <sql-file>` 과 같은 경우에도 해당되는 거겠지요?
입력을 파일로부터 받을 때!

hyunikn replied:
-i 경우는 테스트 해 보았습니다만. 사실 성립하지 않는 경우를 "아직" 찾지 못했습니다. 
코드를 좀 더 꼼꼼이 분석하거나 regression 테스트를 하면 드러날 수도 있습니다.

ctshim replied:
```
[cubrid@ctshim198 databases]$ cat plcsql.sql
create or replace procedure t () as
begin
dbms_output.put_line('hello');
end;
call [change_sp_owner]('t', 'DBA') on class [db_root];

[cubrid@ctshim198 databases]$

```
위와 같이  마지막에 공백을  추가한 후 아래와 같은 명령을 수행하면  assert가 발생합니다. ( `assert(buff[r - 1] == ';');` )
`csql -S -udba testdb -i  plcsql.sql --no-single-line`

hyunikn replied:
감사합니다. 수정해 보겠습니다.

hyunikn commented:
새 구현에서 한 가지 중요한 가정을 하고 있습니다. 
" loaddb의 schema 파일에서 CREATE ... /* pl/csql */  문은 항상 세미콜론으로 끝나며 그 세미콜론 앞에는 공란이 없다."
라는 가정입니다. 
문장의 시작을 CREATE token 의 buffer_pos 로부터 구하는데 있어서 필수적인 가정입니다.

hyunikn committed:
minor comment

vimkim commented on src/parser/csql_lexer.l:
```diff
@@ -1983,74 +2010,76 @@ parser_yyinput_single_line (char *buff, int max_size)
-	    }
-	    end = c;
+           {
+             str_identifier = c;
+             buff[i++] = c;
+             if (i >= max_size)
+               {
+                 return i;
+               }
+           }
+           end = c;
 
-	  do
-	    {
-	      c = pt_nextchar ();
-	      if (c == end)	/* string end */
-		{
-		  str_identifier = '\0';
-		  break;
-		}
+         do
+           {
+             c = pt_nextchar ();
+             if (c == end)     /* string end */
+               {
+                 str_identifier = '\0';
+                 break;
+               }
 
-	      buff[i++] = c;
-	      if (i >= max_size)
-		{
-		  return i;
-		}
+             buff[i++] = c;
+             if (i >= max_size)
+               {
+                 return i;
+               }
 
```
사소한 이슈인데,
tab -> space로 변경된 것을 알려드립니다. 
codestyle.sh가 l, y 파일에는 적용되지 않아서 발생하는 이슈로 보입니다.

hyunikn replied:
아, 복 붙하면서 그리 되었나보네요. 
혼란을 줄 수 있으니 원복 하도록 하겠습니다.

hyunikn committed:
[minor] removed diff caused by tab to spaces change

hyunikn committed:
Merge branch 'feature/plcsql-p1n' into cbrd-25464-loaddb-with-plcsql

hyunikn committed:
let the tokens' csql_yylloc.buffer_pos coincide with the actual position in the file,
and use the positions to get the user sQL texts from the parser->file

hyunikn commented on src/parser/csql_grammar.y:
```diff
@@ -552,7 +553,6 @@ static char *g_plcsql_text;
 
 %}
 
-%initial-action {yybuffer_pos = 0;}
```
yyparse() 가 호출될 때마다 yybuffer_pos 가 0 으로 초기화되면 cubrid loaddb -s 에서 문제를 일으키게 되었습니다. 
parse_one_statement 의 최초 호출과 parser_main 에서 초기화 해주는 것으로 충분합니다.

hyunikn committed:
minor comment updates

hyunikn committed:
minor - removed a comment

hyunikn committed:
exclude comment clause after the PL/CSQL program body

hyunikn committed:
Merge branch 'feature/plcsql-p1n' into cbrd-25464-loaddb-with-plcsql

hyunikn committed:
minor - removed a temporary debugging code
